[chore] Automatically convert tracker links

andrewnicols opened the pull request:
As requeted by @stronk7 in #295

stronk7 reviewed (approved):
Thanks @andrewnicols , I think this makes things easier, specially if at some point our Tracker moves/changes, plus pretty easier to write just the issue numbers.

A few comments (all them optional, related questions):

1. ~I imagine it works, heh. Not sure how to test that. But I trust you~. 
2. Related with the previous point, but unrelated with this issue... do we have any plan to add unit testing to all this stuff?
3. Do we need to add any title/tooltip to those links, or that's not needed?

My side, approving this as is, feel free to merge it!

Ciao 😄 

Edited: Forget about point 1, I previewed it @ netifly (and 3.9.0 release notes).

andrewnicols commented:
> Thanks @andrewnicols , I think this makes things easier, specially if at some point our Tracker moves/changes, plus pretty easier to write just the issue numbers.

Welcome :)

> A few comments (all them optional, related questions):
> 
> 1. ~I imagine it works, heh. Not sure how to test that. But I trust you~.
Addressed :)

> 2. Related with the previous point, but unrelated with this issue... do we have any plan to add unit testing to all this stuff?
I do, and I've been lazy about it for some time. I actually got tests working on this months ago, but then pulled the original functionality. I've added it back and added a couple of basic unit tests for it now.

> 3. Do we need to add any title/tooltip to those links, or that's not needed?
I don't believe there's a way of doing that from Markdown - markdown ins deliberately basic in that regard.

> My side, approving this as is, feel free to merge it!
I'll hold off 24 hours on merging in case you want to reviewi the unit tests one final time, and then I'll self-merge.

Note:
I added a commit to address some missing imports I detected whilst testing this.

stronk7 commented:
Ah, I thought this was pretty standard...
```
My favorite search engine is [Duck Duck Go](https://duckduckgo.com "The best search engine for privacy").
```
My favorite search engine is [Duck Duck Go](https://duckduckgo.com "The best search engine for privacy").

I remember it from the past, including original Markdown (John Gruber) specs. In any case, it's not critical!

Nice to have the tests added and the functionality covered... BTW... is there any reference about how those "elements" (a text, a link) are built? I understand they are sort of concatenated... but then there is the link text being a child of the link element... is that explained anywhere? Just curiosity.

Ciao :-)

andrewnicols commented:
> Ah, I thought this was pretty standard...
> 
> ```
> My favorite search engine is [Duck Duck Go](https://duckduckgo.com "The best search engine for privacy").
> ```
> 
> My favorite search engine is [Duck Duck Go](https://duckduckgo.com).

Ahh - yes. I forgot about those.
I'll land this issue for now, but out of interest, what did you have in mind as a tooltip? We can do it - it's just a case of setting a `title` element next to the URL.

> I remember it from the past, including original Markdown (John Gruber) specs. In any case, it's not critical!

> Nice to have the tests added and the functionality covered... BTW... is there any reference about how those "elements" (a text, a link) are built? I understand they are sort of concatenated... but then there is the link text being a child of the link element... is that explained anywhere? Just curiosity.

They're a part of the MDAST spec (https://github.com/syntax-tree/mdast).

Docusaurus currently uses version 1.6.22 of MDX-remark which uses mdast and adds a few elements to create an MDXAST spec.

Here's a few of the relevant parts of the mdast spec:
- [link](https://github.com/syntax-tree/mdast#link)
- [text](https://github.com/syntax-tree/mdast#text)
- [heading](https://github.com/syntax-tree/mdast#heading)

In our case we're taking a value like:

```
See MDL-12345 for more information
```

Converting it to an AST like:
```
{
    type: 'root',
    children: {
        type: 'paragraph',
        children: [
            {
                type: 'text',
                value: 'See MDL-12345 for more information',
            },
        ],
    },
}
```

We want to convert that `MDL-12345` to a link whose content is MDL-12345, that is:
```
<a href="https://tracker.moodle.org/browse/MDL-12345">MDL-12345</a>
```

That's represented by:
- A fragment of text `See ` (with trailing space)
- A link with:
  - URL: `https://tracker.moodle.org/browse/MDL-12345`
  - Content: `MDL-12345`
- A fragment of text with ` for more information` (with leading space)


So the children become:
```
children: [
    {
        type: 'text',
        value: 'See ',
    },
    {
        type: 'link',
        url: 'https://tracker.moodle.org/browse/MDL-12345',
        children: [
            {
                type: 'text',
                value: 'MDL-12345',
            },
        ],
    },
        type: 'text',
        value: ' for more information',
    },
],
```

If we wanted to add a title (tooltip) to the URL, then we could do so by adding a `title` attribute next to the url:
```
{
    type: 'link',
    url: 'https://tracker.moodle.org/browse/MDL-12345',
    title: 'View the details of MDL-12345 in the Moodle tracker',
    children: [
        {
            type: 'text',
            value: 'MDL-12345',
        },
    ],
},
```

You can see an extended version of that AST in live docs on the mdxjs playground: https://mdxjs.com/playground/
